### PR TITLE
support new targets

### PR DESCRIPTION
**Pull request by hieudion**
We introduce some new targets in this PR
- Score or response, use in NPS
- Recipient, use in NPS
- Custom object
- Custom relationship
- User Event

### CODE REVIEW

File: src/main/java/org/embulk/input/zendesk/clients/ZendeskRestClient.java [hunk 1]
```diff
@@ -127,27 +126,34 @@ public void onGiveup(Exception firstException, Exception lastException)
...
+            if (rateLimiter != null) {
+                rateLimiter.acquire();
+            }
             logger.info(">>> {}{}", request.getURI().getPath(),
                     request.getURI().getQuery() != null ? "?" + request.getURI().getQuery() : "");
-            HttpResponse response = client.execute(request);
-            getRateLimiter(response).acquire();
-            int statusCode = response.getStatusLine().getStatusCode();
+            final HttpResponse response = client.execute(request);
+            if (rateLimiter == null) {
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/clients/ZendeskRestClient.java:149]: Because we acquire `rateLimit` in case it is not null by:
```
            if (rateLimiter != null) {
                rateLimiter.acquire();
            }
```
Then initialize it if it is null, does it mean we don't need to acquire in the first time?

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        return false;
+    }
+
+    @Override
+    public TaskReport execute(final int taskIndex, final RecordImporter recordImporter)
+    {
+        final List<String> paths = getListPathByTarget();
+
+        paths.parallelStream().forEach(path -> StreamSupport.stream(new CustomObjectSpliterator(path, getZendeskRestClient(), task, Exec.isPreview()), Exec.isPreview())
+                .forEach(jsonNode -> recordImporter.addRecord(jsonNode)));
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: `recordImporter::addRecord` is enough

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    private ZendeskRestClient zendeskRestClient;
+
+    public ZendeskCustomObjectService(final ZendeskInputPlugin.PluginTask task)
+    {
+        this.task = task;
+    }
+
+    @VisibleForTesting
+    protected ZendeskRestClient getZendeskRestClient()
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: Please put `protected` method down to the last `public` method. I'd prefer ordering method by its visibility.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    public JsonNode getData(String path, final int page, final boolean isPreview, final long startTime)
+    {
+        if (path.isEmpty()) {
+            path = buildURI(page, startTime);
+        }
+        try {
+            final String response = getZendeskRestClient().doGet(path, task, isPreview);
+            return ZendeskUtils.parseJsonObject(response);
+        }
+        catch (final Exception e) {
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: why should it catch exception here? no checked exception is thrown.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            importDataForIncremental(task, recordImporter, taskReport);
+        }
+        else {
+            importDataForNonIncremental(task, taskIndex, recordImporter);
+        }
+
+        return taskReport;
+    }
+
+    public JsonNode getData(String path, final int page, final boolean isPreview, final long startTime)
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: Naming is confused here, in method naming, what is the different between `getData` and a private `fetchData`?

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        try {
+            final String response = getZendeskRestClient().doGet(path, task, isPreview);
+            return ZendeskUtils.parseJsonObject(response);
+        }
+        catch (final Exception e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    protected URIBuilder getURIBuilderFromHost()
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: I don't see why we need dedicated method here, because all it do only call to the utility method.

File: src/main/java/org/embulk/input/zendesk/utils/ZendeskUtils.java [hunk 7]
```diff
@@ -1,150 +1,88 @@
...
-     * */
-    private static Timestamp getTimestampValue(String value)
+    public static URIBuilder getURIBuilder(final String urlString)
+    {
+        final URI uri;
+        try {
+            uri = new URI(urlString);
+        }
+        catch (final URISyntaxException e) {
+            throw new ConfigException("Login url is invalid format " + urlString);
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/utils/ZendeskUtils.java]: The exception message does not fit with the name of method, method name is general for URI, exception is specific to `login_url` only.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        // For incremental target, we will run in one task but split in multiple threads inside for data deduplication.
+        // Run with incremental will contain duplicated data.
+        ThreadPoolExecutor pool = null;
+        try {
+            final Set<String> knownIds = ConcurrentHashMap.newKeySet();
+            pool = new ThreadPoolExecutor(
+                    10, 100, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>()
+            );
+
+            long resultEndTime = 0;
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: I think we actually don't need that `resultEndTime` local var, just assign `result.get(ZendeskConstants.Field.END_TIME).asLong()` to `startTime` directly at the end.

File: src/main/java/org/embulk/input/zendesk/stream/PagingSpliterator.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+import org.embulk.input.zendesk.clients.ZendeskRestClient;
+import org.embulk.input.zendesk.models.Target;
+import org.embulk.input.zendesk.models.ZendeskException;
+import org.embulk.input.zendesk.utils.ZendeskUtils;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+public abstract class PagingSpliterator<E> implements Spliterator<E>
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/stream/PagingSpliterator.java:8]: I feel we're building an abstract parent class which contains all of common methods for its child classes, so it behaves pretty much similar as utility class. What if more new targets with different endpoints will be used? Does this abstract need to be modified for those new endpoint?

I'd suggest restructure the `PagingSpliterator` and its child classes to bring clear sense.

**Review by trung-huynh — CHANGES_REQUESTED** [review]
I left few comments, still scanning through source code.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+    protected URIBuilder getURIBuilderFromHost()
+    {
+        return ZendeskUtils.getURIBuilder(task.getLoginUrl());
+    }
+
+    private void importDataForIncremental(final ZendeskInputPlugin.PluginTask task, final RecordImporter recordImporter, final TaskReport taskReport)
+    {
+        long startTime = 0;
+        if (!Exec.isPreview() && isSupportIncremental()) {
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: I have few points here:
1. In the original code snippet, `Exec.isPreview` is not checked: https://github.com/treasure-data/embulk-input-zendesk/blob/d669a2c063ecbb17e141254d8be68b7fca00f399/src/main/java/org/embulk/input/zendesk/ZendeskInputPlugin.java#L249-L251, I'm not sure why it brought here, and why `startTime` should not be updated if it is not in preview phase?
2. I think `isSupportIncremental` is redundant, I'd assume the method `importDataForIncremental` is able to call in incremental mode only.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+                    }
+
+                    pool.submit(() -> fetchData(recordJsonNode, task, recordImporter));
+                    recordCount++;
+                    if (Exec.isPreview()) {
+                        return;
+                    }
+                }
+
+                ZendeskNormalServices.logger.info("Fetched '{}' records from start_time '{}'", recordCount, startTime);
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: why don't we just use `logger`?

File: src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        return false;
+    }
+
+    @Override
+    public TaskReport execute(final int taskIndex, final RecordImporter recordImporter)
+    {
+        if (Exec.isPreview()) {
+            JsonNode  jsonNode = mockJsonNode();
+            recordImporter.addRecord(jsonNode.get(0));
+        }
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java:55]: I think it is safe to return `Exec.newTaskReport` here to get rid of `if` `else` block

**Review by trung-huynh — CHANGES_REQUESTED** [review]
I left few more comments.

File: src/main/java/org/embulk/input/zendesk/clients/ZendeskRestClient.java (hunk 1 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/clients/ZendeskRestClient.java:149]: In case we meet some problem, such as rate limit, at the first time the header will not have **x-rate-limit,** which require to init the rate limit. 
So **rateLimiter** will be null, acquire will throw NPE

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 8 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: We need to store last end_time to update config_diff 

>  if (!Exec.isPreview()) {
                storeStartTimeForConfigDiff(taskReport, resultEndTime);
            }

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java (hunk 2 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 5 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: change name fetchData to fetchRecord

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java (hunk 3 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 4 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 6 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: fixed

File: src/main/java/org/embulk/input/zendesk/utils/ZendeskUtils.java (hunk 7 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/utils/ZendeskUtils.java]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 10 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: fixed

File: src/main/java/org/embulk/input/zendesk/stream/PagingSpliterator.java (hunk 9 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/stream/PagingSpliterator.java:8]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java (hunk 3 above)
Comment (trung-huynh) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: hmm, why don't you move `protected` down to `public` instead of changing it to public? Also please scan through source code for other classes too, e.g. https://github.com/treasure-data/embulk-input-zendesk/blob/821f7c9a5a933cc5350842d4ee85f6bce9c15f94/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java#L43-L52 or https://github.com/treasure-data/embulk-input-zendesk/blob/2843e6d771ad38ddb147003335eb1244db1e0f78/src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java#L41-L48.

I'd recommend move it down to public methods instead of changing modifier from `protected` to `public` though.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 8 above)
Comment (trung-huynh) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: Actually we can just use `startTime` as a replacement for `resultEndTime`, what if:

```
diff --git a/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java b/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
index 84921a3..6f51f94 100644
--- a/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
+++ b/src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java
@@ -92,14 +92,12 @@ public abstract class ZendeskNormalServices implements ZendeskService
                     10, 100, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>()
             );
 
-            long resultEndTime = 0;
             while (true) {
                 int recordCount = 0;
 
                 // Page argument isn't used in incremental API so we just set it to 0
                 final JsonNode result = getData("", 0, false, startTime);
                 final Iterator<JsonNode> iterator = ZendeskUtils.getListRecords(result, task.getTarget().getJsonName());
-                resultEndTime = result.get(ZendeskConstants.Field.END_TIME).asLong();
 
                 int numberOfRecords = 0;
                 if (result.has(ZendeskConstants.Field.COUNT)) {
@@ -130,16 +128,15 @@ public abstract class ZendeskNormalServices implements ZendeskService
                 }
 
                 logger.info("Fetched '{}' records from start_time '{}'", recordCount, startTime);
+                startTime = result.get(ZendeskConstants.Field.END_TIME).asLong();
 
                 if (numberOfRecords < ZendeskConstants.Misc.MAXIMUM_RECORDS_INCREMENTAL) {
                     break;
                 }
-
-                startTime = resultEndTime;
             }
 
             if (!Exec.isPreview()) {
-                storeStartTimeForConfigDiff(taskReport, resultEndTime);
+                storeStartTimeForConfigDiff(taskReport, startTime);
             }
         }
         finally {
```

File: src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+                    "        \"size\": 6\n" +
+                    "      },\n" +
+                    "      \"user_id\": \"12312354234\"\n" +
+                    "    }\n" +
+                    "]";
+
+            return new ObjectMapper().readTree(mockData);
+        }
+        catch (IOException ex) {
+            throw new RuntimeException("Error when mock data for guess or preview " + ex.getMessage());
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java]: I'd suggest just said `Can not create sample data {ex.getMessage()}`

File: src/main/java/org/embulk/input/zendesk/RecordImporter.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+import org.embulk.spi.Schema;
+import org.embulk.spi.json.JsonParser;
+import org.embulk.spi.time.Timestamp;
+import org.slf4j.Logger;
+
+import java.util.function.Function;
+
+public class RecordImporter
+{
+    Schema schema;
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/RecordImporter.java]: please reduce the scope to private, we don't need class-modifier here.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        final List<String> paths = path.isEmpty()
+                ? getListPathByTarget()
+                : Collections.singletonList(path);
+
+        for (final String temp : paths) {
+            try {
+                response = Optional.ofNullable(getZendeskRestClient().doGet(temp, task, true));
+
+                // in guessing, break when we have data
+                if (response.isPresent()) {
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java:67]: How can you know it is in `guess` phase? I understand there is a "convention" here that `getData` method of ` ZendeskCustomObjectService` only be called by guess only but I do not prefer the conventional think like that because it might be accidentally called.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 5 above)
Comment (trung-huynh) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: I still think `getData` and `fetchRecord` are quite confused and related to each other, still thinking the better name here to propose.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java (hunk 13 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskUserEventService.java]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 8 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: fixed

File: src/test/java/org/embulk/input/zendesk/clients/TestZendeskRestClient.java [hunk 16]
```diff
@@ -63,10 +62,10 @@
     private PluginTask task = ZendeskTestHelper.getConfigSource("incremental.yml").loadConfig(PluginTask.class);
```
Comment (trung-huynh) [inline src/test/java/org/embulk/input/zendesk/clients/TestZendeskRestClient.java:49]: I know it was here for a while, but please replace deprecated piece of code:
```
private final ExpectedException thrown = ExpectedException.none();
    private final TestingEmbulk embulk = TestingEmbulk.builder()
            .registerPlugin(InputPlugin.class, "zendesk", ZendeskInputPlugin.class)
            .build();
    public ZendeskPluginTestRuntime runtime = new ZendeskPluginTestRuntime();

    @Rule
    @SuppressFBWarnings("URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD")
    public TestRule chain = RuleChain.outerRule(embulk).around(runtime).around(thrown);
```
to
```
@SuppressFBWarnings("URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD")
    @Rule
    public EmbulkTestRuntime embulkTestRuntime = new EmbulkTestRuntime();
```

**Review by trung-huynh — CHANGES_REQUESTED** [review]
Left few comments, still scanning through the whole source code.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+                    pool.submit(() -> fetchRecord(recordJsonNode, task, recordImporter));
+                    recordCount++;
+                    if (Exec.isPreview()) {
+                        return;
+                    }
+                }
+
+                logger.info("Fetched '{}' records from start_time '{}'", recordCount, startTime);
+
+                startTime = result.get(ZendeskConstants.Field.END_TIME).asLong();
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: IIRC `end_time` here is inclusive, `start_time` is also inclusive, will the duplicate happen here because `end_time` will be fetched twice?

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskUserEventService.java [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class TestZendeskUserEventService
```
Comment (trung-huynh) [inline src/test/java/org/embulk/input/zendesk/services/TestZendeskUserEventService.java:30]: I have few concerns here:
* UT only tests the function the class exposes, we don't expose guess, run or preview in `ZendeskUserEventService`, I understand that conventionally `execute` is used in `preview` and `run`, `getData` is used in `guess` phase, but we're designing the new method and we supposed that it could be invoked in any place.
* I'm not 100% sure but is there different test cases naming style in the same plugin, if you look at what you have done before: https://github.com/treasure-data/embulk-input-zendesk/blob/716c8a581a7d0310229d5f8a3687334bf097c306/src/test/java/org/embulk/input/zendesk/TestZendeskInputPlugin.java#L242-L250 and what you just wrote: `testUrlForUserEvent`.

And that concern not only apply for this `UserEventService` but also apply for others.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskService.java [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.embulk.config.TaskReport;
+import org.embulk.input.zendesk.RecordImporter;
+
+public interface ZendeskService
+{
+    boolean isSupportIncremental();
+
+    TaskReport execute(int taskIndex, RecordImporter recordImporter);
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskService.java]: I'd suggest we find the better method name here, in my understanding, `execute` name means we pass the callable function, then executor will execute it, similar as `retry executor`. Our `execute` here will add record into `recordImporter`, should we change to something like `addToRecordImporter` or another name which is less confusing than execute.

**Review by trung-huynh — CHANGES_REQUESTED** [review]
Left few comments, please check carefully because those comments might bring the big changes.

File: src/main/java/org/embulk/input/zendesk/services/ZendeskService.java (hunk 19 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskService.java]: fixed

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskUserEventService.java (hunk 18 above)
Comment (hieudion) [reply src/test/java/org/embulk/input/zendesk/services/TestZendeskUserEventService.java:30]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java (hunk 17 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskNormalServices.java]: It could be an edge case. 
More information: https://develop.zendesk.com/hc/en-us/community/posts/360001646028-Incremental-Ticket-Export-more-than-1000-updates-per-second

File: src/test/java/org/embulk/input/zendesk/clients/TestZendeskRestClient.java (hunk 16 above)
Comment (hieudion) [reply src/test/java/org/embulk/input/zendesk/clients/TestZendeskRestClient.java:49]: fixed

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java (hunk 15 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java:67]: fixed

Comment (trung-huynh) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java:67]: Actually you just update the comment sounds not enough. My point here `getDataFromPath` is assumed only called during `preview` or `guess` process which knows only when dive into code, `isPreview` argument is actually useless.

I understood that `custom object` does not support incremental and `isPreview` must be always `true`. What if we do the `Precondition` check that `isPreview` must be true to execute further that method? Otherwise invoker in the future might call it with `false` value and they see another result.

**Comment by trung-huynh** [comment]
@hieudion because you updated the business which `share the same time` does not exist anymore, should those test cases: https://github.com/treasure-data/embulk-input-zendesk/blob/e3d40efac247026902f8471ea08560a08ec71121/src/test/java/org/embulk/input/zendesk/services/TestZendeskNPSService.java#L67-L88 be removed?

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java [hunk 20]
```diff
@@ -32,14 +39,17 @@
...
     @Before
     public void prepare()
     {
         zendeskRestClient = mock(ZendeskRestClient.class);
+        recordImporter = mock(RecordImporter.class);
     }
 
     @Test
-    public void buildPathWithIncrementalForPreview()
+    public void testBuildPathWithIncrementalForPreview()
```
Comment (trung-huynh) [inline src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: There is no different between `Preview` and `Run` here: https://github.com/treasure-data/embulk-input-zendesk/blob/e3d40efac247026902f8471ea08560a08ec71121/src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java#L83-L90 because it is skipped to check: https://github.com/treasure-data/embulk-input-zendesk/blob/e3d40efac247026902f8471ea08560a08ec71121/src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java#L321

I think it is okay to keep one only.

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java [hunk 21]
```diff
@@ -17,9 +22,11 @@
...
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import java.time.Instant;
 import java.util.Collections;
 
 public class TestZendeskSupportAPIService
```
Comment (trung-huynh) [inline src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: This test class contains many test cases which are in abstract parent class `ZendeskNormalService`, I'd suggest we create another test class for that abstract one to keep it isolated with the subclasses. Subclasses only test what happen in those subclasses.

It will give us the advantage if we implement new services in that future which are similar as support, we just need to add test cases for those services only. Or if we need to update `ZendeskNormalService` we only need to update test cases in that abstract one instead of update other subclasses test cases.

**Review by trung-huynh — COMMENTED** [review]
Left few comments, the logic looks pretty okay for me

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java (hunk 21 above)
Comment (hieudion) [reply src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: fixed

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java (hunk 20 above)
Comment (hieudion) [reply src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: fixed

Comment (trung-huynh) [reply src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: Although it is not 100% related but I would like to spot few points here too:
1. https://github.com/treasure-data/embulk-input-zendesk/blob/273793bceaec3249a2cce9293ef08d693bf7de8f/src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java#L59-L85, different between preview / run is still here
2. Because there is no different between `preview` and `run`, always mention it in test cases are redundant
3. `ZendeskSupportAPIService` only overrides `buildURI`, verify that method only seems enough to me, currently we're mocking the whole data and use its abstract: https://github.com/treasure-data/embulk-input-zendesk/blob/273793bceaec3249a2cce9293ef08d693bf7de8f/src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java#L158, why don't you just follow the one you wrote in another class: https://github.com/treasure-data/embulk-input-zendesk/blob/273793bceaec3249a2cce9293ef08d693bf7de8f/src/test/java/org/embulk/input/zendesk/services/TestZendeskNPSService.java#L34-L42?

File: src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java [hunk 22]
```diff
@@ -39,25 +37,27 @@ public void prepare()
...
         setup("incremental.yml");
         loadData("data/tickets.json");
-        setupTestAndVerifyURL(expectURL, 0, true);
+        String url = zendeskSupportAPIService.buildURI(0, 0);
+        assertEquals(expectURL, url);
     }
 
     @Test
-    public void buildPathWithNonIncrementalForPreview()
+    public void testBuildPathWithNonIncrementalFor()
```
Comment (trung-huynh) [inline src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java]: Duplicated as https://github.com/treasure-data/embulk-input-zendesk/blob/091c1094c4ca0a9c87615249d75f15af83a35d01/src/test/java/org/embulk/input/zendesk/services/TestZendeskSupportAPIService.java#L102-L110?

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        return false;
+    }
+
+    @Override
+    public TaskReport addRecordToImporter(final int taskIndex, final RecordImporter recordImporter)
+    {
+        final List<String> paths = getListPathByTarget();
+
+        paths.parallelStream().forEach(path -> StreamSupport.stream(new CustomObjectSpliterator(path, getZendeskRestClient(), task, Exec.isPreview()), Exec.isPreview())
```
Comment (trung-huynh) [inline src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: is it running parallel only in `preview` mode?

**Review by trung-huynh — COMMENTED** [review]
left comment

File: src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java (hunk 23 above)
Comment (hieudion) [reply src/main/java/org/embulk/input/zendesk/services/ZendeskCustomObjectService.java]: It is a mistake. Should be running parallel in run mode only. Fixed

**Review by trung-huynh — APPROVED** [review]
LGTM

**Comment by hieudion** [comment]
Thank @trung-huynh for your helps. I am going to merger it.
